<a href="https://colab.research.google.com/github/karlmanalo/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_211_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 1*

---

# Regression 1

## Assignment

You'll use another **New York City** real estate dataset. 

But now you'll **predict how much it costs to rent an apartment**, instead of how much it costs to buy a condo.

The data comes from renthop.com, an apartment listing website.

- [ ] Look at the data. Choose a feature, and plot its relationship with the target.
- [ ] Use scikit-learn for linear regression with one feature. You can follow the [5-step process from Jake VanderPlas](https://jakevdp.github.io/PythonDataScienceHandbook/05.02-introducing-scikit-learn.html#Basics-of-the-API).
- [ ] Define a function to make new predictions and explain the model coefficient.
- [ ] Organize and comment your code.

> [Do Not Copy-Paste.](https://docs.google.com/document/d/1ubOw9B3Hfip27hF2ZFnW3a3z9xAgrUDRReOEo-FHCVs/edit) You must type each of these exercises in, manually. If you copy and paste, you might as well not even do them. The point of these exercises is to train your hands, your brain, and your mind in how to read, write, and see code. If you copy-paste, you are cheating yourself out of the effectiveness of the lessons.

If your **Plotly** visualizations aren't working:
- You must have JavaScript enabled in your browser
- You probably want to use Chrome or Firefox
- You may need to turn off ad blockers
- [If you're using Jupyter Lab locally, you need to install some "extensions"](https://plot.ly/python/getting-started/#jupyterlab-support-python-35)

## Stretch Goals
- [ ] Do linear regression with two or more features.
- [ ] Read [The Discovery of Statistical Regression](https://priceonomics.com/the-discovery-of-statistical-regression/)
- [ ] Read [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf), Chapter 2.1: What Is Statistical Learning?

In [0]:
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [0]:
# Read New York City apartment rental listing data
import pandas as pd
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

In [0]:
# Remove outliers: 
# the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= 1375) & (df['price'] <= 15500) & 
        (df['latitude'] >=40.57) & (df['latitude'] < 40.99) &
        (df['longitude'] >= -74.1) & (df['longitude'] <= -73.38)]

In [4]:
# Exploring data

df.head()

bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0
2        1.0         1  ...                 0                    0
3        1.0         1  ...                 0                    0
4        1.0         4  ...                 0                    0

[5 rows x 34 columns]

In [5]:
# Exploring numeric data

df.describe()

bathrooms      bedrooms  ...  wheelchair_access  common_outdoor_space
count  48818.000000  48818.000000  ...       48818.000000          48818.000000
mean       1.201790      1.537158  ...           0.027224              0.026240
std        0.470707      1.106078  ...           0.162736              0.159851
min        0.000000      0.000000  ...           0.000000              0.000000
25%        1.000000      1.000000  ...           0.000000              0.000000
50%        1.000000      1.000000  ...           0.000000              0.000000
75%        1.000000      2.000000  ...           0.000000              0.000000
max       10.000000      8.000000  ...           1.000000              1.000000

[8 rows x 29 columns]

In [0]:
import plotly.express as px

In [7]:
# Visually exploring data to find trendline with highest R2 value

px.scatter(df, x='bathrooms', y='price', trendline='ols')

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning:

pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.



In [0]:
# Importing Linear Regression model from SKLearn

from sklearn.linear_model import LinearRegression

In [0]:
# Instantiating model

model = LinearRegression()

In [0]:
# Defining x and y variables for training set

features = ['bathrooms']
target = ['price']

X_train = df[features]
y_train = df[target]

In [32]:
# Fitting model to training set

model.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [33]:
# Applying model to new data

bathrooms = 3
x_test = [[ bathrooms ]]

y_pred = model.predict(x_test)
y_pred

array([[8207.02772077]])

In [34]:
# Defining function to predict price based on number of bathrooms in apartment.

def predict(bathrooms):
  y_pred = model.predict([[bathrooms]])
  estimate = y_pred[0]
  print(int(estimate), "is our estimated monthly rent for a NYC apartment with",
        int(bathrooms), "bathrooms.")

predict(3)

8207 is our estimated monthly rent for a NYC apartment with 3 bathrooms.


In [35]:
print(model.coef_)

[[2573.37439508]]


The model coefficient for this linear regression model implies that, for every additional bathroom in a New York City apartment, the monthly rent for said apartment increases by $2,573.37. This is the slope of the line represented by our linear regression model.

In [14]:
px.scatter(df, x=df['bedrooms'], y=df['price'], trendline='ols')

# Stretch Goals:

In [0]:
mlr = LinearRegression()

In [0]:
features = ['bathrooms', 'bedrooms']
target = ['price']

X_train = df[features]
y_train = df[target]

In [17]:
mlr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [18]:
bathrooms = 3
bedrooms = 3
x_test = [[ bathrooms, bedrooms ]]

y_pred = mlr.predict(x_test)
y_pred

array([[7917.58758932]])

In [19]:
def predict(bathrooms, bedrooms):
  y_pred = mlr.predict([[bathrooms, bedrooms]])
  estimate = y_pred[0]
  print(int(estimate), "is our estimated monthly rent for a NYC apartment with",
        int(bedrooms), "bedrooms and", int(bathrooms), "bathrooms.")
  
predict(3, 3)

7917 is our estimated monthly rent for a NYC apartment with 3 bedrooms and 3 bathrooms.


# Read both of the suggested reading materials.

In [0]:
# px.scatter(df, x=df['latitude'], y=df['price'])

In [0]:
# px.scatter(df, x=df['longitude'], y=df['price'])

In [0]:
# from sklearn.linear_model import ElasticNet

In [0]:
# model = ElasticNet()

In [24]:
'''features = ['latitude']
target = ['price']

X_train = df[features]
y_train = df[target]'''

"features = ['latitude']\ntarget = ['price']\n\nX_train = df[features]\ny_train = df[target]"

In [0]:
# model.fit(X_train, y_train)

In [26]:
'''latitude = 40.6
x_test = [[ latitude ]]

y_pred = model.predict(x_test)
y_pred'''

'latitude = 40.6\nx_test = [[ latitude ]]\n\ny_pred = model.predict(x_test)\ny_pred'

In [0]:
# model.coef_

In [0]:
# model.intercept_